### PR TITLE
feat(smart-contracts): remove UDT mint on mainnet purchases

### LINKED ISSUE #10706: Stop minting from mainnet
The protocol behaves differently on mainnet and on other networks. On mainnet new UDT are minted while on other networks they are distributed.
Let's change this so that tokens are also distributed from the premine on mainnet.

As part of this change, we will need to redeploy the mainnet contract. This means that we need to include a mechanism in the `PublicLock` contract (and in its new version) so that when they are upgraded, they "point" to the new Unlock contract. Additionnaly the new Unlock needs to "accept" locks deployed from the older contract as "valid" locks.

### PR DESCRIPTION
**Pull request by clemsos**
# Description

Use the premint on mainnet to distribute tokens instead of minting new tokens

# Issues

Fixes #10706
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/Unlock.sol [hunk 1]
```diff
@@ -452,24 +452,13 @@ contract Unlock is UnlockInitializable, UnlockOwnable {
...
-              2 /
-              grossNetworkProduct;
-          }
+          // we distribute tokens using asymptotic curve between 0 and 0.5
+          // maxTokens = IMintableERC20(udt).balanceOf(address(this)).mul((valueInETH / grossNetworkProduct) / (2 + 2 * valueInETH / grossNetworkProduct));
+          uint maxTokens = (IMintableERC20(udt).balanceOf(
+              address(this)
+            ) * valueInETH) /
+            (2 + (2 * valueInETH) / grossNetworkProduct) /
+            grossNetworkProduct;
```
Comment (julien51) [inline smart-contracts/contracts/Unlock.sol:461]: I wonder if there is a way to simplify this equation?

Comment (clemsos) [reply smart-contracts/contracts/Unlock.sol:461]: that may mean change how UDT distribution works . Should write a proposal about tokenomics first as it is a change in the protocol itself?

Comment (julien51) [reply smart-contracts/contracts/Unlock.sol:461]: No no, I want to keep the same formula but have the code be more legible!

Comment (clemsos) [reply smart-contracts/contracts/Unlock.sol]: Ok. A simple thing is to reuse the `balance` variable below. We are hitting "stack too deep" so can not really add more named vars :/

**Review by julien51 — APPROVED** [review]

**Pull request merged**